bvm/ballerina-runtime/src/main/java/io/ballerina/runtime/internal/regexp/RegExpFactory.java

```diff
@@ -0,0 +1,189 @@
+        return new RegExpAtomQuantifier(reAtom, quantifier);
+    }
+
+    public static RegExpLiteralCharOrEscape createReLiteralCharOrEscape(BString charOrEscape) {
+        return new RegExpLiteralCharOrEscape(charOrEscape.getValue());
+    }
+
+    public static RegExpCharacterClass createReCharacterClass(BString characterClassStart, BString negation,
+                                                              RegExpCharSet reCharSet, BString characterClassEnd) {
+        return new RegExpCharacterClass(characterClassStart.getValue(), negation.getValue(), reCharSet,
+                characterClassEnd.getValue());
+    }
+
+    public static RegExpCharSet createReCharSet(ArrayValue charSet) {
+        return new RegExpCharSet(charSet);
+    }
+
+    public static RegExpCharSetRange createReCharSetRange(BString lhsCharSetAtom, BString dash,
+                                                          BString rhsCharSetAtom) {
+        return new RegExpCharSetRange(lhsCharSetAtom.getValue(), dash.getValue(), rhsCharSetAtom.getValue());
+    }
+
+    public static RegExpCapturingGroup createReCapturingGroup(BString openParen, RegExpFlagExpression flagExpr,
+                                                              RegExpDisjunction reDisjunction, BString closeParen) {
+        return new RegExpCapturingGroup(openParen.getValue(), flagExpr, reDisjunction, closeParen.getValue());
+    }
+
+    public static RegExpFlagExpression createReFlagExpression(BString questionMark, RegExpFlagOnOff flagsOnOff,
+                                                              BString colon) {
+        return new RegExpFlagExpression(questionMark.getValue(), flagsOnOff, colon.getValue());
+    }
+
+    public static RegExpFlagOnOff createReFlagOnOff(BString flags) {
+        return new RegExpFlagOnOff(flags.getValue());
+    }
+
+    public static RegExpQuantifier createReQuantifier(BString quantifier, BString nonGreedyChar) {
+        return new RegExpQuantifier(quantifier.getValue(), nonGreedyChar.getValue());
+    }
+
+    public static RegExpValue parse(String regExpStr) {
+        try {
+            CharReader charReader = CharReader.from(regExpStr);
+            TokenReader tokenReader = new TokenReader(new TreeTraverser(charReader));
+            TreeBuilder treeBuilder = new TreeBuilder(tokenReader);
+            return treeBuilder.parse();
+        } catch (BallerinaException e) {
+            throw ErrorCreator.createError(StringUtils.fromString("Failed to parse regular expression: " +
+                    e.getMessage()));
+        }
+    }
+
+    public static RegExpDisjunction translateRegExpConstructs(RegExpValue regExpValue) {
+        RegExpDisjunction disjunction = regExpValue.getRegExpDisjunction();
+        for (Object s : disjunction.getRegExpSeqList()) {
+            if (s == null) {
+                break;
+            }
+            if (!(s instanceof RegExpSequence)) {
```

Comment:
Can we use node type checks instead?

Reply:
Check all places

Reply:
We don't have specific runtime types to identify these. Hence need to use `instanceOf` check